test/interpreter_functional/test_suites/run_pipeline/helpers.js

```diff
@@ -0,0 +1,89 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import expect from 'expect.js';
+import { delay } from 'bluebird';
+
+export const expectExpressionProvider = ({ getService, updateBaselines }) => {
+  const browser = getService('browser');
+  const screenshot = getService('screenshots');
+  const snapshots = getService('snapshots');
+  const log = getService('log');
+
+  return (name, expression, context = {}, initialContext = {}) => {
+    log.debug(`executing expression ${expression}`);
+    const steps = expression.split('|'); // todo: we should actually use interpreter parser and get the ast
+    let responsePromise;
+
+    const handler = {
+      toReturn: async result => {
+        const pipelineResponse = await handler.getResponse();
+        expect(pipelineResponse).to.eql(result);
+      },
+      getResponse: () => {
+        if (!responsePromise) responsePromise = handler.runExpression();
+        return responsePromise;
+      },
+      runExpression: async (step, stepContext) => {
+        log.debug(`running expression ${step || expression}`);
+        const promise = browser.executeAsync((expression, context, initialContext, done) => {
+          window.runPipeline(expression, context, initialContext).then(result => {
+            done(result);
+          });
+        }, step || expression, stepContext || context, initialContext);
+        return await promise;
+      },
+      stepsToMatchSnapshot: async () => {
+        let lastResponse;
+        for (let i = 0; i < steps.length; i++) {
+          const step = steps[i];
+          lastResponse = await handler.runExpression(step, lastResponse);
+          const diff = await snapshots.compareAgainstBaseline(name + i, lastResponse, updateBaselines);
+          expect(diff).to.be.lessThan(0.05);
+        }
+        if (!responsePromise) {
+          responsePromise = new Promise(resolve => {
+            resolve(lastResponse);
+          });
+        }
+        return handler;
+      },
+      toMatchSnapshot: async () => {
+        const pipelineResponse = await handler.getResponse();
+        await snapshots.compareAgainstBaseline(name, pipelineResponse, updateBaselines);
+        return handler;
+      },
+      toMatchScreenshot: async () => {
+        const pipelineResponse = await handler.getResponse();
+        await browser.executeAsync((context, done) => {
+          window.renderPipelineResponse(context).then(result => {
+            done(result);
+          });
+        }, pipelineResponse);
+        await delay(1000);
+
+        const percentDifference = await screenshot.compareAgainstBaseline(name, updateBaselines);
+        expect(percentDifference).to.be.lessThan(0.05);
+        return handler;
+      }
+    };
+
+    return handler;
+  };
+};
```

Comment:
nit: Once the dust settles, some jsdoc comments might be nice for the handler (or docs in the README)